scipy/sparse/linalg/_eigen/_svds.py

```diff
@@ -338,49 +333,60 @@ def matmat_XH_X(x):
     elif solver == 'arpack' or solver is None:
         if v0 is None and not rs_was_None:
             v0 = random_state.uniform(size=(min(A.shape),))
-        eigvals, eigvec = eigsh(XH_X, k=k, tol=tol ** 2, maxiter=maxiter,
+        _, eigvec = eigsh(XH_X, k=k, tol=tol ** 2, maxiter=maxiter,
                                 ncv=ncv, which=which, v0=v0)
 
-    # Gramian matrices have real non-negative eigenvalues.
-    eigvals = np.maximum(eigvals.real, 0)
-
-    # Use the sophisticated detection of small eigenvalues from pinvh.
-    t = eigvec.dtype.char.lower()
-    factor = {'f': 1E3, 'd': 1E6}
-    cond = factor[t] * np.finfo(t).eps
-    cutoff = cond * np.max(eigvals)
-
-    # Get a mask indicating which eigenpairs are not degenerately tiny,
-    # and create the re-ordered array of thresholded singular values.
-    above_cutoff = (eigvals > cutoff)
-    nlarge = above_cutoff.sum()
-    nsmall = k - nlarge
-    slarge = np.sqrt(eigvals[above_cutoff])
-    s = np.zeros_like(eigvals)
-    s[:nlarge] = slarge
+    eigvec, _ = np.linalg.qr(eigvec)
+    u = X_matmat(eigvec)
     if not return_singular_vectors:
-        return np.sort(s)
+        s = svd(u, compute_uv=False, overwrite_a=True)
+        return s[::-1]
 
-    if n > m:
-        vlarge = eigvec[:, above_cutoff]
-        ularge = (X_matmat(vlarge) / slarge
-                  if return_singular_vectors != 'vh' else None)
-        vhlarge = _herm(vlarge)
+    # compute the left singular vectors of X and update the right ones
+    # accordingly
+    u, s, vh = svd(u, full_matrices=False, overwrite_a=True)
+    u = u[:, ::-1]
+    s = s[::-1]
+    vh = vh[::-1]
+
+    jobu = return_singular_vectors in {True, 'u'}
+    jobv = return_singular_vectors in {True, 'vh'}
+
+    if tol is None or tol <= 0.0:
+        res_tol = np.sqrt(1e-15) * max(A.shape)
     else:
-        ularge = eigvec[:, above_cutoff]
-        vhlarge = (_herm(X_matmat(ularge) / slarge)
-                   if return_singular_vectors != 'u' else None)
-
-    u = (_augmented_orthonormal_cols(ularge, nsmall, random_state)
-         if ularge is not None else None)
-    vh = (_augmented_orthonormal_rows(vhlarge, nsmall, random_state)
-          if vhlarge is not None else None)
-
-    indexes_sorted = np.argsort(s)
-    s = s[indexes_sorted]
-    if u is not None:
-        u = u[:, indexes_sorted]
-    if vh is not None:
-        vh = vh[indexes_sorted]
+        res_tol = tol
+
+    if transpose:
+        u_tmp = eigvec @ _herm(vh) if jobu else None
+        if jobv:
+            # if singular values are very small, the right singular vectors
+            # computed above may be inaccurate, so we have to re-compute
+            # them as the eigenvectors of X_XH
+            eigvals, eigvec = lobpcg(X_XH, u, tol=res_tol**2, maxiter=maxiter,
```

Comment:
@evgueni-ovtchinnikov are eigvals actually being used? If not, use _ instead

Reply:
ok